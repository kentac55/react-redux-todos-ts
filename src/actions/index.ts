import actionCreatorFactory from 'typescript-fsa'
import { AsyncOpResult, Todo, VisibilityFilter } from '../types'

const actionCreator = actionCreatorFactory()

type AddTodoProp = {
  text: string
}
export const addTodo = actionCreator.async<
  AddTodoProp,
  AsyncOpResult<Todo>,
  Error
>('AddTodo')

type ToggleTodoProp = {
  todo: Todo
}
export const toggleTodo = actionCreator.async<
  ToggleTodoProp,
  AsyncOpResult<Todo>,
  Error
>('ToggleTodo')

export const setVisibilityFilter = actionCreator<VisibilityFilter>(
  'SetVisibilityFilter'
)

type GetTodoProp = {
  id: number
}
export const getTodo = actionCreator.async<
  GetTodoProp,
  AsyncOpResult<Todo>,
  Error
>('GetTodo')

type GetTodosProp = void
export const getTodos = actionCreator.async<
  GetTodosProp,
  AsyncOpResult<Todo[]>,
  Error
>('GetTogos')

type InitProp = void
export const initApp = actionCreator.async<
  InitProp,
  AsyncOpResult<Todo[]>,
  Error
>('InitApp')
