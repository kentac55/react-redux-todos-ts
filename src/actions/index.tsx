import actionCreatorFactory from 'typescript-fsa'
import { VisibilityFilterKinds } from '../types'

const actionCreator = actionCreatorFactory()
export const addTodoAction = actionCreator<string>('ADD')
export const toggleTodoAction = actionCreator<number>('TOGGLE')

export const setVisibilityFilterAction = actionCreator<VisibilityFilterKinds>(
  'SET'
)
