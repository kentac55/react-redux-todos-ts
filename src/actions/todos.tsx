let nextTodoId = 0
export const addTodoAction = (text: string) =>
  ({
    type: 'ADD',
    id: nextTodoId++,
    text,
  } as const)

export const toggleTodoAction = (id: number) =>
  ({
    type: 'TOGGLE',
    id,
  } as const)
