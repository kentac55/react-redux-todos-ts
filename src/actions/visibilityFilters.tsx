import { VisibilityFilterKinds } from '../types'

export const setVisibilityFilterAction = (filter: VisibilityFilterKinds) =>
  ({
    type: 'SET_VISIBILITY_FILTER',
    filter,
  } as const)
