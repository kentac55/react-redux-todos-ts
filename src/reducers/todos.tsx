import { Reducer } from 'redux'
import { Todo, TodoAction } from '../types'

const todos: Reducer<Todo[], TodoAction> = (
  state: Todo[] = [],
  action: TodoAction
): Todo[] => {
  switch (action.type) {
    case 'ADD':
      return [
        ...state,
        {
          id: action.id,
          text: action.text,
          completed: false,
        },
      ]
    case 'TOGGLE':
      return state.map(
        (todo: Todo): Todo => {
          return todo.id === action.id
            ? {
                ...todo,
                completed: !todo.completed,
              }
            : todo
        }
      )
    default:
      return state
  }
}

export default todos
