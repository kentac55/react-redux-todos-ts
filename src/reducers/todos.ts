import { reducerWithInitialState } from 'typescript-fsa-reducers'
import { Todo } from '../types'
import { getId } from '../repository'
import { addTodoAction, toggleTodoAction } from '../actions'

export const todoReducer = reducerWithInitialState([] as Todo[])
  .case(addTodoAction, (state, text) => {
    return [...state, { id: getId(), text, completed: false }]
  })
  .case(toggleTodoAction, (state, id) => {
    return state.map(todo => {
      return todo.id === id
        ? {
            ...todo,
            completed: !todo.completed,
          }
        : todo
    })
  })
