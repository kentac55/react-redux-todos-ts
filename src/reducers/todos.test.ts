import { Action } from 'redux'
import { todoReducer, TodoState } from './todos'
import { addTodo, initApp, getTodos, toggleTodo } from '../actions'

const todos = [
  { id: 0, text: 'hoge', completed: false },
  { id: 1, text: 'fuga', completed: true },
  { id: 2, text: 'piyo', completed: false },
]
const id = 4
const text = 'test'
const todo = { id, text, completed: false }

const zeroState = {
  todos: [],
}

const someState = {
  todos,
}

const serverError = { type: 'Internal', message: null } as const

it('should return initial state', () => {
  expect(todoReducer(undefined, { type: undefined })).toEqual(zeroState)
})

it.each<[TodoState, Action, TodoState]>([
  [
    zeroState,
    initApp.done({
      params: undefined,
      result: { type: 'Ok', contents: todos },
    }),
    someState,
  ],
  [
    zeroState,
    initApp.done({ params: undefined, result: serverError }),
    zeroState,
  ],
  [
    someState,
    addTodo.done({
      params: { text },
      result: { type: 'Ok', contents: todo },
    }),
    { todos: [...todos, todo] },
  ],
  [
    someState,
    addTodo.done({ params: { text }, result: serverError }),
    someState,
  ],
  [
    someState,
    toggleTodo.done({
      params: { todo: todos[0] },
      result: {
        type: 'Ok',
        contents: { ...todos[0], completed: !todos[0].completed },
      },
    }),
    {
      todos: [
        { ...todos[0], completed: !todos[0].completed },
        todos[1],
        todos[2],
      ],
    },
  ],
  [
    someState,
    toggleTodo.done({ params: { todo }, result: serverError }),
    someState,
  ],
  [
    zeroState,
    getTodos.done({
      params: undefined,
      result: { type: 'Ok', contents: todos },
    }),
    someState,
  ],
  [
    zeroState,
    getTodos.done({ params: undefined, result: serverError }),
    zeroState,
  ],
])('should handle %s', (initState, action, expectedState) => {
  expect(todoReducer(initState, action)).toStrictEqual(expectedState)
  expect(todoReducer(initState, action)).toMatchSnapshot()
})
