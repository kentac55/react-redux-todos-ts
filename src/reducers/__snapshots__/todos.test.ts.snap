// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should handle { todos: [] } 1`] = `
Object {
  "todos": Array [
    Object {
      "completed": false,
      "id": 0,
      "text": "hoge",
    },
    Object {
      "completed": true,
      "id": 1,
      "text": "fuga",
    },
    Object {
      "completed": false,
      "id": 2,
      "text": "piyo",
    },
  ],
}
`;

exports[`should handle { todos: [] } 2`] = `
Object {
  "todos": Array [],
}
`;

exports[`should handle { todos: [] } 3`] = `
Object {
  "todos": Array [
    Object {
      "completed": false,
      "id": 0,
      "text": "hoge",
    },
    Object {
      "completed": true,
      "id": 1,
      "text": "fuga",
    },
    Object {
      "completed": false,
      "id": 2,
      "text": "piyo",
    },
  ],
}
`;

exports[`should handle { todos: [] } 4`] = `
Object {
  "todos": Array [],
}
`;

exports[`should handle { todos: [Array] } 1`] = `
Object {
  "todos": Array [
    Object {
      "completed": false,
      "id": 0,
      "text": "hoge",
    },
    Object {
      "completed": true,
      "id": 1,
      "text": "fuga",
    },
    Object {
      "completed": false,
      "id": 2,
      "text": "piyo",
    },
    Object {
      "completed": false,
      "id": 4,
      "text": "test",
    },
  ],
}
`;

exports[`should handle { todos: [Array] } 2`] = `
Object {
  "todos": Array [
    Object {
      "completed": false,
      "id": 0,
      "text": "hoge",
    },
    Object {
      "completed": true,
      "id": 1,
      "text": "fuga",
    },
    Object {
      "completed": false,
      "id": 2,
      "text": "piyo",
    },
  ],
}
`;

exports[`should handle { todos: [Array] } 3`] = `
Object {
  "todos": Array [
    Object {
      "completed": true,
      "id": 0,
      "text": "hoge",
    },
    Object {
      "completed": true,
      "id": 1,
      "text": "fuga",
    },
    Object {
      "completed": false,
      "id": 2,
      "text": "piyo",
    },
  ],
}
`;

exports[`should handle { todos: [Array] } 4`] = `
Object {
  "todos": Array [
    Object {
      "completed": false,
      "id": 0,
      "text": "hoge",
    },
    Object {
      "completed": true,
      "id": 1,
      "text": "fuga",
    },
    Object {
      "completed": false,
      "id": 2,
      "text": "piyo",
    },
  ],
}
`;
